library/L1_Peripheral/stm32f10x/pwm.hpp

```diff
@@ -0,0 +1,446 @@
+
+    inline static const PwmPin_t kB9 = {
+      .pin       = pb9,
+      .registers = TIM4,
+      .channel   = 4,
+      .id        = SystemController::Peripherals::kTimer4
+    };
+  };  // Timer
+
+  /// Constructor for a STM32F10x PWM channel.
+  ///
+  /// @param timer - Reference to a const timer description for this
+  ///        instance of the PWM driver.
+  explicit constexpr Pwm(const PwmPin_t & timer) : pwm_pin_(timer) {}
+
+  void ConfigureFrequency(units::frequency::hertz_t frequency) override
+  {
+    static constexpr auto kUpdateGeneration = bit::MaskFromRange(0);
+
+    auto & system                   = SystemController::GetPlatformController();
+    const auto kPeripheralFrequency = system.GetClockRate(pwm_pin_.id);
+
+    auto period      = 1 / frequency;
+    uint32_t product = kPeripheralFrequency * period;
+
+    pwm_pin_.registers->PSC = GetPrescalarValue(product);
+    pwm_pin_.registers->ARR =
+      static_cast<uint16_t>(product / (pwm_pin_.registers->PSC + 1));
+
+    bit::Register(&pwm_pin_.registers->EGR).Set(kUpdateGeneration).Save();
+  }
+
+  void SetDutyCycle(float duty_cycle) override
+  {
+    // Clamp the duty cycle to make sure it's in the right range
+    const float kClampedDutyCycle = std::clamp(duty_cycle, 0.0f, 1.0f);
+
+    switch (pwm_pin_.channel)
+    {
+    case 1:
+      pwm_pin_.registers->CCR1 =
+        static_cast<uint16_t>(kClampedDutyCycle * pwm_pin_.registers->ARR);
+      break;
+
+    case 2:
+      pwm_pin_.registers->CCR2 =
+        static_cast<uint16_t>(kClampedDutyCycle * pwm_pin_.registers->ARR);
+      break;
+
+    case 3:
+      pwm_pin_.registers->CCR3 =
+        static_cast<uint16_t>(kClampedDutyCycle * pwm_pin_.registers->ARR);
+      break;
+
+    case 4:
+      pwm_pin_.registers->CCR4 =
+        static_cast<uint16_t>(kClampedDutyCycle * pwm_pin_.registers->ARR);
+      break;
+    }
+  }
+
+  float GetDutyCycle() override
+  {
+    return
+      static_cast<float>(pwm_pin_.registers->CCR1 / pwm_pin_.registers->ARR);
+  }
+
+  void ModuleEnable(bool enable = true) override
+  {
+    // Preload enable
+    static constexpr auto kPreloadEnableLow  = bit::MaskFromRange(3);
+    static constexpr auto kPreloadEnableHigh = bit::MaskFromRange(11);
+
+    // Output channel enable
+    static constexpr auto kChannel1OutputEnable = bit::MaskFromRange(0);
+    static constexpr auto kChannel2OutputEnable = bit::MaskFromRange(4);
+    static constexpr auto kChannel3OutputEnable = bit::MaskFromRange(8);
+    static constexpr auto kChannel4OutputEnable = bit::MaskFromRange(12);
+
+    // Main Output Enable (Timer 1 only)
+    static constexpr auto kMainOutputEnable = bit::MaskFromRange(15);
+
+    // Timer enable
+    static constexpr auto kTimerEnable = bit::MaskFromRange(0);
+
+    if (enable)
+    {
+      // Preload enable
+      switch (pwm_pin_.channel)
+      {
+      case 1:
+        bit::Register(&pwm_pin_.registers->CCMR1)
+          .Set(kPreloadEnableLow)
+          .Save();
```

Comment:
```suggestion
        bit::Register(&pwm_pin_.registers->CCMR1)
          .Set(kPreloadEnableLow)
          .Set(kChannel1OutputEnable)
          . // etc
          .Save();
```

You can chain these together. The bit::Register() class is efficient if you chain a bunch of modifications together and then at the end `Save()` the final result. That way the computation can occur mostly in registers, sometimes in cache, worst case using RAM. But the benefit is that it doesn't have to access all the way into peripherals for every statement. Also chaining these calls is actually constexpr, so a lot of this turns into a single load instruction, but the code still looks very expressive.

This also eliminates the multiple uses of switch case where you may be able to get it down to just 1.

Reply:
Unfortunately this won't work because all the registers are different. One switch affects CCMRx, the other does CCER, and then BDTR and CR1. Unless I'm misunderstanding how the chaining works.

Reply:
So one option to get rid of the switches would be to store witch CCRx register each PwmPin_t object uses. CCMRx is little more tricky since we'd have to store the bit mask as well but that way we could get rid of the channel entry in the struct too. So overall we'd add one entry to the struct and get rid of all switches. Thoughts?

Reply:
We would actually still need the switch for CCER at that point unless we add another bit mask entry.

Reply:
Ah ok I missed that. Still doesn't explain why there are two switch case blocks if both of them use the same channel variable. You should group both statements under the same case.